Workflow cannot be saved

When trying to saving a workflow, it gets the following error. This was happening with the workflow from the migration test VM which did not get all the arrows (after correcting them manually):

Date/time: 2020-06-10 13:47:03  
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:77.0) Gecko/20100101 Firefox/77.0  
User IP: 192.168.91.1  
Request URI: /kitodo/pages/workflowEdit.jsf  
Ajax request: Yes  
Status code: 500  
Exception type: class org.camunda.bpm.model.xml.ModelParseException  
Exception message: SAXException while parsing input stream  
Exception UUID:  
Stack trace:

```
    org.camunda.bpm.model.xml.ModelParseException: SAXException while parsing input stream
    	at org.camunda.bpm.model.xml.impl.util.DomUtil.parseInputStream(DomUtil.java:245)
    	at org.camunda.bpm.model.xml.impl.parser.AbstractModelParser.parseModelFromStream(AbstractModelParser.java:98)
    	at org.camunda.bpm.model.bpmn.impl.BpmnParser.parseModelFromStream(BpmnParser.java:65)
    	at org.camunda.bpm.model.bpmn.Bpmn.doReadModelFromInputStream(Bpmn.java:249)
    	at org.camunda.bpm.model.bpmn.Bpmn.readModelFromStream(Bpmn.java:131)
    	at org.kitodo.production.workflow.model.Reader.loadProcess(Reader.java:118)
    	at org.kitodo.production.workflow.model.Reader.<init>(Reader.java:67)
    	at org.kitodo.production.workflow.model.Converter.<init>(Converter.java:63)
    	at org.kitodo.production.forms.WorkflowForm.hasMultipleSructureTreeConfiguration(WorkflowForm.java:175)
    	at org.kitodo.production.forms.WorkflowForm.saveAndRedirect(WorkflowForm.java:146)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at org.apache.el.parser.AstValue.invoke(AstValue.java:247)
    	at org.apache.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:267)
    	at org.jboss.weld.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
    	at org.jboss.weld.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
    	at org.apache.myfaces.view.facelets.el.ContextAwareTagMethodExpression.invoke(ContextAwareTagMethodExpression.java:96)
    	at org.apache.myfaces.application.ActionListenerImpl.processAction(ActionListenerImpl.java:74)
    	at org.springframework.faces.webflow.FlowActionListener.processAction(FlowActionListener.java:71)
    	at org.springframework.faces.model.SelectionTrackingActionListener.processAction(SelectionTrackingActionListener.java:64)
    	at javax.faces.component.UICommand.broadcast(UICommand.java:120)
    	at javax.faces.component.UIViewRoot._broadcastAll(UIViewRoot.java:1255)
    	at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:420)
    	at javax.faces.component.UIViewRoot._process(UIViewRoot.java:1741)
    	at javax.faces.component.UIViewRoot.processApplication(UIViewRoot.java:935)
    	at org.apache.myfaces.lifecycle.InvokeApplicationExecutor.execute(InvokeApplicationExecutor.java:42)
    	at org.apache.myfaces.lifecycle.LifecycleImpl.executePhase(LifecycleImpl.java:195)
    	at org.apache.myfaces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:142)
    	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:204)
    	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292)
    	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
    	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
    	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
    	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
    	at org.kitodo.production.servletfilter.EncodingFilter.doFilter(EncodingFilter.java:69)
    	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
    	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
    	at org.kitodo.production.security.SecurityObjectAccessFilter.doFilter(SecurityObjectAccessFilter.java:56)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
    	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.session.ConcurrentSessionFilter.doFilter(ConcurrentSessionFilter.java:155)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:66)
    	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
    	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
    	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
    	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
    	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
    	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
    	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
    	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
    	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:213)
    	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
    	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:494)
    	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
    	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
    	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
    	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
    	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:522)
    	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1095)
    	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:672)
    	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1520)
    	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1476)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    	at java.lang.Thread.run(Thread.java:748)
    Caused by: org.xml.sax.SAXException: Error: URI=null Line=42: cvc-complex-type.4: Attribut 'sourceRef' muss in Element 'bpmn2:sequenceFlow' vorkommen.
    	at org.camunda.bpm.model.xml.impl.util.DomUtil$DomErrorHandler.error(DomUtil.java:202)
    	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.error(ErrorHandlerWrapper.java:137)
    	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:396)
    	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
    	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:284)
    	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator$XSIErrorReporter.reportError(XMLSchemaValidator.java:453)
    	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.reportSchemaError(XMLSchemaValidator.java:3231)
    	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.addDefaultAttributes(XMLSchemaValidator.java:2937)
    	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.handleStartElement(XMLSchemaValidator.java:2055)
    	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.emptyElement(XMLSchemaValidator.java:761)
    	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:351)
    	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2784)
    	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602)
    	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
    	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505)
    	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:842)
    	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:771)
    	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
    	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:243)
    	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
    	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:121)
    	at org.camunda.bpm.model.xml.impl.util.DomUtil.parseInputStream(DomUtil.java:240)
    	... 90 more
```

Database to reproduce this: [dump.sql](https://gist.github.com/matthias-ronge/0149beb6410bf8bc23cab42164d970b1)

Steps to reproduce: Go to system » migrate. Click “Create workflow”, select all projects, proceed, select first line from database (with 5 processes). In the window that appears, choose “create new workflow”. Do required changes to workflow (add missing arrows, assign roles to steps), click “save”.

--- a/Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java
+++ b/Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java
@@ -303,14 +303,15 @@ public class MigrationForm extends BaseForm {
         TasksToWorkflowConverter templateConverter = new TasksToWorkflowConverter();
         List<Task> processTasks = blueprintProcess.getTasks();
         processTasks.sort(Comparator.comparingInt(Task::getOrdering));
+        String workflowTitle = "ChangeME_"+ Helper.generateRandomString(3);
 
         try {
-            templateConverter.convertTasksToWorkflowFile("ChangeME", processTasks);
+            templateConverter.convertTasksToWorkflowFile(workflowTitle, processTasks);
         } catch (IOException e) {
             Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
         }
 
-        workflowToUse = new Workflow("ChangeME");
+        workflowToUse = new Workflow(workflowTitle);
         workflowToUse.setClient(blueprintProcess.getProject().getClient());
         workflowToUse.setStatus(WorkflowStatus.DRAFT);
         workflowToUse.getTemplates().add(null);
@@ -337,6 +338,7 @@ public class MigrationForm extends BaseForm {
         if (Objects.nonNull(workflowId) && workflowId != 0) {
             // showPopup for Template
             try {
+                workflowToUse = ServiceManager.getWorkflowService().getById(workflowId);
                 createTemplates();
             } catch (DAOException e) {
                 Helper.setErrorMessage(ERROR_READING, new Object[] {ObjectType.TEMPLATE.getTranslationSingular() },

--- a/Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java
+++ b/Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java
@@ -303,7 +303,7 @@ public class MigrationForm extends BaseForm {
         TasksToWorkflowConverter templateConverter = new TasksToWorkflowConverter();
         List<Task> processTasks = blueprintProcess.getTasks();
         processTasks.sort(Comparator.comparingInt(Task::getOrdering));
-        String workflowTitle = "ChangeME_"+ Helper.generateRandomString(3);
+        String workflowTitle = "ChangeME_" + Helper.generateRandomString(3);
 
         try {
             templateConverter.convertTasksToWorkflowFile(workflowTitle, processTasks);

--- a/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
+++ b/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
@@ -166,6 +166,25 @@ public class WorkflowForm extends BaseForm {
         }
     }
 
+    /**
+     * Cancel Workflow creation.
+     * @return redirectPath
+     */
+    public String cancel() {
+        if(migration){
+            try {
+                ServiceManager.getWorkflowService().remove(workflow);
+            } catch (DataException e) {
+                Helper.setErrorMessage(ERROR_DELETING, new Object[] {this.workflow.getTitle(), e.getMessage() }, logger,
+                        e);
+                return this.stayOnCurrentPage;
+            }
+            return MIGRATION_FORM_PATH;
+        }
+
+        return "projects?keepPagination=true";
+    }
+
     private boolean hasMultipleSructureTreeConfiguration() throws WorkflowException, IOException {
         Map<String, String> requestParameterMap = FacesContext.getCurrentInstance().getExternalContext()
                 .getRequestParameterMap();

--- a/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
+++ b/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
@@ -171,12 +171,12 @@ public class WorkflowForm extends BaseForm {
      * @return redirectPath
      */
     public String cancel() {
-        if(migration){
+        if (migration) {
             try {
                 ServiceManager.getWorkflowService().remove(workflow);
             } catch (DataException e) {
                 Helper.setErrorMessage(ERROR_DELETING, new Object[] {this.workflow.getTitle(), e.getMessage() }, logger,
-                        e);
+                    e);
                 return this.stayOnCurrentPage;
             }
             return MIGRATION_FORM_PATH;

--- a/Kitodo/src/main/java/org/kitodo/production/migration/TasksToWorkflowConverter.java
+++ b/Kitodo/src/main/java/org/kitodo/production/migration/TasksToWorkflowConverter.java
@@ -75,7 +75,7 @@ public class TasksToWorkflowConverter {
             int diagramOrder = i + 1;
             if (i == 0) {
                 diagram.append(
-                    XmlGenerator.generateTask(tasks.get(i), "StartEvent_1", "Task_" + diagramOrder, diagramOrder));
+                    XmlGenerator.generateTask(tasks.get(i), "StartEvent_1", diagramOrder));
             } else {
                 diagram.append(XmlGenerator.generateTask(tasks.get(i), diagramOrder));
             }

--- a/Kitodo/src/main/java/org/kitodo/production/migration/TasksToWorkflowConverter.java
+++ b/Kitodo/src/main/java/org/kitodo/production/migration/TasksToWorkflowConverter.java
@@ -71,10 +71,11 @@ public class TasksToWorkflowConverter {
     }
 
     private void convertTasksToXmlWorkflow(StringBuilder diagram, List<Task> tasks) {
-        for (int i = 0; i<tasks.size(); i++) {
-            int diagramOrder = i+1;
-            if (i==0) {
-                diagram.append(XmlGenerator.generateTask(tasks.get(i), "StartEvent_1", "Task_" + diagramOrder, diagramOrder));
+        for (int i = 0; i < tasks.size(); i++) {
+            int diagramOrder = i + 1;
+            if (i == 0) {
+                diagram.append(
+                    XmlGenerator.generateTask(tasks.get(i), "StartEvent_1", "Task_" + diagramOrder, diagramOrder));
             } else {
                 diagram.append(XmlGenerator.generateTask(tasks.get(i), diagramOrder));
             }
@@ -84,7 +85,7 @@ public class TasksToWorkflowConverter {
     private void convertTasksToWorkflowCoordinates(StringBuilder diagram, List<Task> tasks) {
         int diagramStart = 498;
 
-        for (int i = 1; i<=tasks.size(); i++) {
+        for (int i = 1; i <= tasks.size(); i++) {
             diagram.append(XmlGenerator.generateTaskShape(i, diagramStart));
             diagramStart = diagramStart + BOX_DISTANCE;
         }

--- a/Kitodo/src/main/java/org/kitodo/production/migration/TasksToWorkflowConverter.java
+++ b/Kitodo/src/main/java/org/kitodo/production/migration/TasksToWorkflowConverter.java
@@ -71,11 +71,12 @@ public class TasksToWorkflowConverter {
     }
 
     private void convertTasksToXmlWorkflow(StringBuilder diagram, List<Task> tasks) {
-        for (Task task : tasks) {
-            if (task.getOrdering().equals(1)) {
-                diagram.append(XmlGenerator.generateTask(task, "StartEvent_1", "Task_" + task.getOrdering()));
+        for (int i = 0; i<tasks.size(); i++) {
+            int diagramOrder = i+1;
+            if (i==0) {
+                diagram.append(XmlGenerator.generateTask(tasks.get(i), "StartEvent_1", "Task_" + diagramOrder, diagramOrder));
             } else {
-                diagram.append(XmlGenerator.generateTask(task));
+                diagram.append(XmlGenerator.generateTask(tasks.get(i), diagramOrder));
             }
         }
     }
@@ -83,8 +84,8 @@ public class TasksToWorkflowConverter {
     private void convertTasksToWorkflowCoordinates(StringBuilder diagram, List<Task> tasks) {
         int diagramStart = 498;
 
-        for (Task task : tasks) {
-            diagram.append(XmlGenerator.generateTaskShape(task.getOrdering(), diagramStart));
+        for (int i = 1; i<=tasks.size(); i++) {
+            diagram.append(XmlGenerator.generateTaskShape(i, diagramStart));
             diagramStart = diagramStart + BOX_DISTANCE;
         }
 

--- a/Kitodo/src/main/java/org/kitodo/production/migration/XmlGenerator.java
+++ b/Kitodo/src/main/java/org/kitodo/production/migration/XmlGenerator.java
@@ -37,7 +37,7 @@ class XmlGenerator {
      * @return generated task
      */
     static String generateTask(Task task, int order) {
-        return generateTask(task, "Task_" + (order - 1), "Task_" + order, order);
+        return generateTask(task, "Task_" + (order - 1), order);
     }
 
     /**
@@ -47,13 +47,10 @@ class XmlGenerator {
      *            for generating
      * @param sourceReference
      *            for sequence flow
-     * @param targetReference
-     *            for sequence flow
      * @return generated task
      */
-    static String generateTask(Task task, String sourceReference, String targetReference, int ordering) {
+    static String generateTask(Task task, String sourceReference, int ordering) {
         StringBuilder taskBuilder = new StringBuilder();
-
         openTask(taskBuilder, task);
 
         taskBuilder.append("id=\"Task_");
@@ -96,7 +93,7 @@ class XmlGenerator {
 
         generateSequences(taskBuilder, ordering);
         closeTask(taskBuilder, task);
-        generateSequenceFlow(taskBuilder, ordering, sourceReference, targetReference);
+        generateSequenceFlow(taskBuilder, ordering, sourceReference);
 
         return taskBuilder.toString();
     }
@@ -149,14 +146,13 @@ class XmlGenerator {
         taskBuilder.append("</bpmn2:outgoing>\n");
     }
 
-    private static void generateSequenceFlow(StringBuilder taskBuilder, Integer ordering, String sourceReference,
-            String targetReference) {
+    private static void generateSequenceFlow(StringBuilder taskBuilder, Integer ordering, String sourceReference) {
         taskBuilder.append("        <bpmn2:sequenceFlow id=\"SequenceFlow_");
         taskBuilder.append(ordering);
         taskBuilder.append("\" sourceRef=\"");
         taskBuilder.append(sourceReference);
         taskBuilder.append("\" targetRef=\"");
-        taskBuilder.append(targetReference);
+        taskBuilder.append("Task_");
         taskBuilder.append(SLASH_END_LINE);
     }
 

--- a/Kitodo/src/main/java/org/kitodo/production/migration/XmlGenerator.java
+++ b/Kitodo/src/main/java/org/kitodo/production/migration/XmlGenerator.java
@@ -59,7 +59,6 @@ class XmlGenerator {
         taskBuilder.append("id=\"Task_");
         taskBuilder.append(ordering);
         taskBuilder.append(QUOTES);
-
         taskBuilder.append("name=\"");
         taskBuilder.append(task.getTitle());
         taskBuilder.append(QUOTES);
@@ -89,12 +88,10 @@ class XmlGenerator {
             taskBuilder.deleteCharAt(taskBuilder.length() - 1);
             taskBuilder.append(QUOTES);
         }
-
         if (StringUtils.isNotBlank(task.getScriptName()) || StringUtils.isNotBlank(task.getScriptPath())) {
             generateTemplateTaskProperty(taskBuilder, "scriptName", task.getScriptName());
             generateTemplateTaskProperty(taskBuilder, "scriptPath", task.getScriptPath());
         }
-
         taskBuilder.append(">\n");
 
         generateSequences(taskBuilder, ordering);

--- a/Kitodo/src/main/java/org/kitodo/production/migration/XmlGenerator.java
+++ b/Kitodo/src/main/java/org/kitodo/production/migration/XmlGenerator.java
@@ -14,6 +14,7 @@ package org.kitodo.production.migration;
 import org.apache.commons.lang3.StringUtils;
 import org.kitodo.data.database.beans.Role;
 import org.kitodo.data.database.beans.Task;
+import org.kitodo.data.database.enums.TaskStatus;
 
 class XmlGenerator {
 
@@ -35,8 +36,8 @@ class XmlGenerator {
      *            for generating
      * @return generated task
      */
-    static String generateTask(Task task) {
-        return generateTask(task, "Task_" + (task.getOrdering() - 1), "Task_" + task.getOrdering());
+    static String generateTask(Task task, int order) {
+        return generateTask(task, "Task_" + (order - 1), "Task_" + order, order);
     }
 
     /**
@@ -50,13 +51,13 @@ class XmlGenerator {
      *            for sequence flow
      * @return generated task
      */
-    static String generateTask(Task task, String sourceReference, String targetReference) {
+    static String generateTask(Task task, String sourceReference, String targetReference, int ordering) {
         StringBuilder taskBuilder = new StringBuilder();
 
         openTask(taskBuilder, task);
 
         taskBuilder.append("id=\"Task_");
-        taskBuilder.append(task.getOrdering());
+        taskBuilder.append(ordering);
         taskBuilder.append(QUOTES);
 
         taskBuilder.append("name=\"");
@@ -64,7 +65,9 @@ class XmlGenerator {
         taskBuilder.append(QUOTES);
 
         generateTemplateTaskProperty(taskBuilder, "editType", task.getEditType().getValue());
-        generateTemplateTaskProperty(taskBuilder, "processingStatus", task.getProcessingStatus().getValue());
+        generateTemplateTaskProperty(taskBuilder, "processingStatus",
+            task.getProcessingStatus().equals(TaskStatus.DONE) ? TaskStatus.DONE.getValue()
+                    : TaskStatus.LOCKED.getValue());
         generateTemplateTaskProperty(taskBuilder, "concurrent", task.isConcurrent());
         generateTemplateTaskProperty(taskBuilder, "typeMetadata", task.isTypeMetadata());
         generateTemplateTaskProperty(taskBuilder, "separateStructure", task.isSeparateStructure());
@@ -94,9 +97,9 @@ class XmlGenerator {
 
         taskBuilder.append(">\n");
 
-        generateSequences(taskBuilder, task.getOrdering());
+        generateSequences(taskBuilder, ordering);
         closeTask(taskBuilder, task);
-        generateSequenceFlow(taskBuilder, task.getOrdering(), sourceReference, targetReference);
+        generateSequenceFlow(taskBuilder, ordering, sourceReference, targetReference);
 
         return taskBuilder.toString();
     }

--- a/Kitodo/src/main/webapp/pages/workflowEdit.xhtml
+++ b/Kitodo/src/main/webapp/pages/workflowEdit.xhtml
@@ -74,9 +74,8 @@
                          actionListener="#{WorkflowForm.setSaveDisabled(false)}"
                          update="save"
                          style="display:none;"/>
-        <p:button value="#{isViewMode ? msgs.close : msgs.cancel}"
-                  onclick="setConfirmUnload(false);"
-                  outcome="projects?keepPagination=true"
+        <p:commandButton value="#{isViewMode ? msgs.close : msgs.cancel}"
+                  action="#{WorkflowForm.cancel()}"
                   icon="fa fa-times fa-lg"
                   iconPos="right"
                   styleClass="#{isViewMode ? 'primary' : 'secondary'}"/>

--- a/Kitodo/src/test/java/org/kitodo/production/migration/XmlGeneratorIT.java
+++ b/Kitodo/src/test/java/org/kitodo/production/migration/XmlGeneratorIT.java
@@ -63,7 +63,7 @@ public class XmlGeneratorIT {
     public void shouldGenerateTaskWithSourceTarget() {
         Task task = template.getTasks().stream().filter(x -> x.getOrdering().equals(1)).findAny().orElse(null);
 
-        String taskString = XmlGenerator.generateTask(task, "StartEvent_1", "Task_1", task.getOrdering());
+        String taskString = XmlGenerator.generateTask(task, "StartEvent_1", task.getOrdering());
         String expected = "        <bpmn2:task id=\"Task_1\" name=\"Finished\" template:editType=\"3\" "
                 + "template:processingStatus=\"3\" template:concurrent=\"false\" template:typeMetadata=\"false\" "
                 + "template:separateStructure=\"false\" template:typeAutomatic=\"false\" template:typeExportDMS=\"false\" "

--- a/Kitodo/src/test/java/org/kitodo/production/migration/XmlGeneratorIT.java
+++ b/Kitodo/src/test/java/org/kitodo/production/migration/XmlGeneratorIT.java
@@ -43,7 +43,7 @@ public class XmlGeneratorIT {
     public void shouldGenerateTaskWithoutSourceTarget() {
         Task task = template.getTasks().stream().filter(x -> x.getOrdering().equals(2)).findAny().orElse(null);
 
-        String taskString = XmlGenerator.generateTask(task);
+        String taskString = XmlGenerator.generateTask(task, task.getOrdering());
         String expected = "        <bpmn2:scriptTask id=\"Task_2\" name=\"Blocking\" template:editType=\"1\" "
                 + "template:processingStatus=\"3\" template:concurrent=\"false\" template:typeMetadata=\"false\" "
                 + "template:separateStructure=\"false\" template:typeAutomatic=\"false\" template:typeExportDMS=\"false\" "
@@ -63,7 +63,7 @@ public class XmlGeneratorIT {
     public void shouldGenerateTaskWithSourceTarget() {
         Task task = template.getTasks().stream().filter(x -> x.getOrdering().equals(1)).findAny().orElse(null);
 
-        String taskString = XmlGenerator.generateTask(task, "StartEvent_1", "Task_1");
+        String taskString = XmlGenerator.generateTask(task, "StartEvent_1", "Task_1", task.getOrdering());
         String expected = "        <bpmn2:task id=\"Task_1\" name=\"Finished\" template:editType=\"3\" "
                 + "template:processingStatus=\"3\" template:concurrent=\"false\" template:typeMetadata=\"false\" "
                 + "template:separateStructure=\"false\" template:typeAutomatic=\"false\" template:typeExportDMS=\"false\" "

--- a/Kitodo/src/test/java/org/kitodo/selenium/MigrationST.java
+++ b/Kitodo/src/test/java/org/kitodo/selenium/MigrationST.java
@@ -14,6 +14,7 @@ package org.kitodo.selenium;
 import static org.awaitility.Awaitility.await;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
 
 import java.io.File;
 import java.net.URI;
@@ -63,7 +64,7 @@ public class MigrationST extends BaseTestSelenium {
         assertEquals("Finished, Closed, Progress, Open, Locked", systemPage.getAggregatedTasks(2));
         WorkflowEditPage workflowEditPage = systemPage.createNewWorkflow();
         workflowEditPage.changeWorkflowStatusToActive();
-        assertEquals("ChangeME", workflowEditPage.getWorkflowTitle());
+        assertTrue(workflowEditPage.getWorkflowTitle().contains("ChangeME"));
         workflowEditPage.changeWorkflowTitle("migrationWorkflow");
         systemPage = workflowEditPage.saveForMigration();
         String newTemplateTitle = "newTemplate";

